README.md
<p align="center">
	<h1 align="center">CreativeSizzle.Redirect</h1>
</p>

<p align="center">
	<em>Manage all your HTTP redirects with an easy to use GUI. This is an essential SEO plugin.</em>
</p>

<p align="center">
	<img src="https://badgen.net/packagist/php/creative-sizzle/wn-redirect-plugin">
	<img src="https://badgen.net/packagist/license/creative-sizzle/wn-redirect-plugin">
	<img src="https://badgen.net/packagist/v/creative-sizzle/wn-redirect-plugin/latest">
	<img src="https://badgen.net/badge/cms/Winter%20CMS">
	<img src="https://badgen.net/badge/type/plugin">
</p>

## The #1 Redirect plugin for Winter CMS

This is the best Redirect-plugin for Winter CMS. With this plugin installed you can manage redirects directly from Winter CMS' beautiful interface. Many webmasters and SEO specialists use redirects to optimise their website for search engines. This plugin allows you to manage such redirects with a nice and user-friendly interface.

## History

This plugin was originally build in 2016 by Alwin Drenth a Software Engineer at Van der Let & Partners.
As of 2018 this plugin is re-distributed to the October CMS Marketplace with vendor name Vdlp.Redirect (formerly known as Adrenth.Redirect).
As of 2022 this plugin is re-distributed to the Winter CMS Marketplace with vendor name CreativeSizzle.Redirect (formerly Vdlp.Redirect).

The Redirect plugin will now be maintained by Creative Sizzle and You (the open source community).

## What does this plugin offer?

This plugin adds a 'Redirects' section to the main menu of Winter CMS. This plugin has a unique and fast matching algorithm to match your redirects before your website is being rendered.

## Features

* **Quick** matching algorithm
* A **test** utility for redirects
* Matching using **placeholders** (dynamic paths)
* Matching using **regular expressions**
* **Exact** path matching
* **Importing** and **exporting** redirect rules
* **Schedule** redirects (e.g. active for 2 months)
* Redirect to **external** URLs
* Redirect to **internal** CMS pages
* Redirect to relative or absolute URLs
* Redirect **log**
* **Categorize** redirects
* **Statistics**
    * Hits per redirect
    * Popular redirects per month (top 10)
    * Popular crawlers per month (top 10)
    * Number of redirects per month
    * And more...
* Multilingual ***(Need help translating!***
* Supports MySQL, SQLite and Postgres
* HTTP status codes 301, 302, 303, 404, 410
* Caching

## Supported database platforms

* MySQL
* Postgres
* SQLite

## Requirements

* Winter CMS 1.1 or higher.
* PHP version 7.4 or higher.
* PHP extensions: `ext-curl` and `ext-json`.

## Supported HTTP status codes

* `HTTP/1.1 301 Moved Permanently`
* `HTTP/1.1 302 Found`
* `HTTP/1.1 303 See Other`
* `HTTP/1.1 404 Not Found`
* `HTTP/1.1 410 Gone`

## Supported HTTP request methods

* `GET`
* `POST`
* `HEAD`

## Performance

All redirects are stored in the database and will be automatically "published" to a file which the internal redirect mechanism uses to determine if a certain request needs to be redirected. This is way faster than querying a database.

This plugin is designed to be fast and should have no negative effect on the performance of your website.

To gain maximum performance with this plugin:

* Enable redirect caching using a "in-memory" caching method (see Caching).
* Maintain your redirects frequently to keep the number of redirects as low as possible.
* Try to use placeholders to keep your number of redirect low (less redirects is better performance).

## Caching

If your website has a lot of redirects it is recommended to enable redirect caching. You can enable redirect caching in the settings panel of this plugin.

Only cache drivers which support tagged cache are supported. So driver `file` and `database` are not supported. For this plugin database and file caching do not increase performance, but can actually have a negative influence on performance. So it is recommended to use an in-memory caching solution like `memcached` or `redis`.

### How caching works

If caching is enabled (and supported) every request which is handled by this plugin will be cached. It will be stored with tag `CreativeSizzle.Redirect`.

When you modify a redirect all redirect cache will be invalidated automatically. It is also possible to manually clear the cache using the 'Clear cache' button in the Backend.

## Placeholders

This plugin makes advantage of the `symfony/routing` package. So if you need more info on how to make placeholder requirements for your redirection URLs, please go to: https://symfony.com/doc/current/components/routing/introduction.html#usage

## Contribution

Please feel free to [contribute](https://github.com/creative-sizzle/wn-redirect-plugin) to this awesome plugin.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [4.0.0] - TBD
* Fork plugin to work with Winter CMS.

## [3.0.2] - 2022-02-20

* Change version constraint for composer/installers.

## [3.0.1] - 2022-02-20

* Add support for regular expression matches in target path.

## [3.0.0] - 2022-02-19

* Drop support for October CMS 1.1 and lower.
* Minimum required PHP version is now 7.4
* Redirect to relative paths is now enabled by default.
* Add database relation to system request logs.
* When updating to `3.0.0` the table `vdlp_redirect_redirect_logs` will be removed due to schema changes.
* Improvements to redirect request logs.
* Improvements to redirect conditions logic.
* Improvements to the statistics page.
* Redirect Settings:
  * Relative paths setting is now **enabled** by default.
  * Testlab (Beta) is now **disabled** by default.
  * Redirect logging is now **disabled** by default.
  * Redirect statistics is now **disabled** by default.

## [2.6.0]

* Update plugin dependencies.

## [2.5.13]

* Fix database error when cache is being cleared before installation of plugin.

## [2.5.12]

* Fix strpos() type error.

## [2.5.11]

* Remove CMS support check.
* Fix bad use of import.

## [2.5.10]

* Add PHP 8.0 version constraint.
* Add composer/installers package.

## [2.5.9]

* Fix import in Plugin file.

## [2.5.8]

* Improve redirect caching management (revised).

## [2.5.7]

* Improve redirect caching management.

## [2.5.6]

* Prevent connection exception when accessing settings in CLI mode.

## [2.5.5]

* Suppress logging when redirect rules file is empty.

## [2.5.4]

* Add support for symfony/stopwatch:^5.0 (version 4.0 is still supported).
* Update Spanish language (thanks to Juan David M).
* Hide button "From Request log" when request logging is disabled.

## [2.5.3]

* Improve / fixes redirect rule caching (thanks to Eric Pfeiffer).
* Update Spanish language (thanks to Juan David M).
* Update Language files (help wanted!).

## [2.5.2]

* Fix bug that causes re-writing the redirect rules file when hits are updated.

## [2.5.1]

* Fixes issues with redirect rules file not being present.

## [2.5.0]

* Add support for using relative paths.

## [2.4.1]

* Add Redirect Extensions promo page.

## [2.4.0]

* Skip requests with header "X-Requested-With: XMLHttpRequest".

## [2.3.2]

* Improve error handling in plugin migration process.

## [2.3.1]

* Fix SQLSTATE[42S22] error when installing plugin.

## [2.3.0]

* Add new Redirect options:
    * Ignore Case
    * Ignore Trailing Slashes
* Fix date field timezone issue (scheduled tab).
* Fix warning dialog position (when scheduled redirect is not active).
* Minor translation improvements (en, nl).

## [2.2.0]

* Add "Cache-Control: no-store" header. This will prevent (modern) web browsers to cache the redirects. Very convenient when testing your redirects.
* Add extra tab "Event logs" to Redirect update page. This tab shows a list with the related event logs of the redirect.
* UI improvements.

## [2.1.1]

* Update CHANGELOG.

## [2.1.0]

* Improve exception handling #52.
* Add support for league/csv:9.0+.
* Improve caching mechanism #54.
* Suppress cache flush log message.
* Skip sparkline routes from being processed.

## [2.0.2]

* Force type of creativesizzle.redirect::log_redirect_changes #53.
* Apply config check to prevent log redirect changes #53.
* Convert database column types (char to varchar) #51.

## [2.0.1]

* Fix Middleware not being invoked in newer PHP versions.

## [2.0.0]

* Drop support for PHP 7.0, only supports PHP 7.1.3+.
* Most of the classes are made final. For extending use October CMS proposed solutions.
* Auto-redirect creation for CMS/Static pages has been removed from this plugin.
* The following events have been removed:
    * `creativesizzle.redirects.changed`
    * `creativesizzle.redirect.beforeRedirectSave`
    * `creativesizzle.redirect.beforeRedirectUpdate`
    * `creativesizzle.redirect.afterRedirectUpdate`
* New events:
    * `creativesizzle.redirect.changed`
    * `creativesizzle.redirect.changed`


lang/sv/lang.php
<?php

return [
    'plugin' => [
        'name' => 'Ompekning',
        'description' => 'Hantera ompekningar',
    ],
    'permission' => [
        'access_redirects' => [
            'label' => 'Hantera ompekningar',
            'tab' => 'Ompekningar',
        ],
    ],
    'navigation' => [
        'menu_label' => 'Ompekningar',
        'menu_description' => 'Hantera ompekningar',
    ],
    'settings' => [
        'menu_label' => 'Redirects',
        'menu_description' => 'Manage settings for Redirects.',
        'logging_enabled_label' => 'Log redirect events',
        'logging_enabled_comment' => 'Store redirect events in the database.',
        'statistics_enabled_label' => 'Gather statistics',
        'statistics_enabled_comment' => 'Gather statistics of redirected requests to get more insight.',
        'test_lab_enabled_label' => 'TestLab (beta)',
        'test_lab_enabled_comment' => 'TestLab allows you to mass test your redirects.',
        'caching_enabled_label' => 'Caching of redirects (advanced)',
        'caching_enabled_comment' => 'Improves the redirect engine when having a lot of redirects. CAUTION: Cache driver `file` and `database` are NOT supported. Recommended driver is `memcached` or a similar "in-memory" caching driver.',
        'relative_paths_enabled_label' => 'Use relative paths',
        'relative_paths_enabled_command' => 'The redirect engine will generate relative paths instead of absolute paths.',
    ],
    'redirect' => [
        'redirect' => 'Ompekning',
        'from_url' => 'Anropsadress',
        'from_url_placeholder' => '/source/path',
        'from_url_comment' => 'Anropsadressen att matcha',
        'from_scheme' => 'Source scheme',
        'from_scheme_comment' => 'Force match on scheme. If HTTP is selected <u>http://domain.com/path</u> will match and <u>https://domain.com/path</u> does not match.',
        'to_url' => 'Måladress',
        'to_url_placeholder' => '/absolute/path, relative/path eller http://target.url',
        'to_url_comment' => 'Måladressen eller url:en att ompeka till',
        'to_url_required_if' => 'Måladressen är obligatorisk',
        'to_scheme' => 'Target scheme',
        'to_scheme_comment' => 'Target scheme will be forced to HTTP or HTTPS or choose AUTOMATIC to use the default scheme of the website.',
        'scheme_auto' => 'Automatic',
        'cms_page_required_if' => 'Ange en CMS-adress att ompeka till',
        'static_page_required_if' => 'Ange en Sidor-sida att ompeka till',
        'match_type' => 'Matcha typ',
        'exact' => 'Exakt',
        'placeholders' => 'Platshållare',
        'regex' => 'Regular expression',
        'target_type' => 'Måltyp',
        'target_type_none' => 'Not applicable',
        'target_type_path_or_url' => 'Mål eller url',
        'target_type_cms_page' => 'CMS-sida',
        'target_type_static_page' => 'Sidor-sida',
        'status_code' => 'HTTP-statuskod',
        'sort_order' => 'Sorteringsordning',
        'requirements' => 'Krav',
        'requirements_comment' => 'Ange krav för varje platshållare',
        'placeholder' => 'Platshållare',
        'placeholder_comment' => 'Platshållarens namn (inklusive måsvingar) som hittas i anropsadressfältet. Ex. {category} eller {id}',
        'requirement' => 'Krav',
        'requirement_comment' => 'Ange kraven i regular expression-syntax. Ex. [0-9]+ eller [a-zA-Z]+.',
        'requirements_prompt' => 'Lägg till krav',
        'replacement' => 'Ersättning',
        'replacement_comment' => 'Ange en valfritt ersättningsvärde för denna platshållare. Den matchade platshållaren kommer att ersättas av detta värde i måladressen',
        'permanent' => '301 - Permanent',
        'temporary' => '302 - Tillfällig',
        'see_other' => '303 - See Other',
        'not_found' => '404 - Ofunnen',
        'gone' => '410 - Gone',
        'enabled' => 'Aktiv',
        'none' => 'none',
        'enabled_comment' => 'Slå på för att aktivera ompekningen',
        'priority' => 'Prioritet',
        'hits' => 'Träffar',
        'return_to_redirects' => 'Återvänd till ompekningar',
        'return_to_categories' => 'Återvänd till kategorier',
        'delete_confirm' => 'Är du säker?',
        'created_at' => 'Skapad',
        'modified_at' => 'Ändrad',
        'system_tip' => 'Systemskapad ompekning',
        'user_tip' => 'Andvändargenererad ompekning',
        'type' => 'Typ',
        'category' => 'Kategori',
        'categories' => 'Kategorier',
        'description' => 'Description',
        'name' => 'Namn',
        'date_time' => 'Date & Time',
        'date' => 'Date',
        'truncate_confirm' => 'Are you sure you want to delete ALL records?',
        'truncating' => 'Deleting...',
        'warning' => 'Warning',
        'cache_warning' => 'You have enabled caching but your caching driver is not supported. Redirects will not be cached.',
        'general_confirm' => 'Are you sure you want to do this?',
        'sparkline_30d' => 'Hits (30d)',
        'has_hits' => 'Has hits',
        'minimum_hits' => 'Minimum # hits',
        'ignore_query_parameters' => 'Ignore query parameters',
        'ignore_query_parameters_comment' => 'The redirect engine will ignore all query parameters.',
        'ignore_case' => 'Ignore case',
        'ignore_case_comment' => 'The redirect engine will do case-insensitive matching.',
        'ignore_trailing_slash' => 'Ignore trailing slash',
        'ignore_trailing_slash_comment' => 'The redirect engine will ignore trailing slashes.',
        'last_used_at' => 'Användes senast',
        'updated_at' => 'Updated at',
        'invalid_regex' => 'Invalid regular expression.',
    ],
    'list' => [
        'no_records' => 'Det finns inga ompekningar i denna vy',
        'switch_is_enabled' => 'Aktiva',
        'switch_system' => 'Systemompekningar',
    ],
    'scheduling' => [
        'from_date' => 'Aktiv från',
        'from_date_comment' => 'Det datum då denna ompekning blir aktiv. Kan lämnas tomt',
        'to_date' => 'Aktiv tills',
        'to_date_comment' => 'Det datum då demma ompekning blir inaktiv. Kan lämas tomt',
        'scheduling_comment' => 'Här kan du ange ett tidsspann då denna ompekning ska vara aktiv. Alla typer av datumkombinationer är möjliga',
        'not_active_warning' => 'Redirect is not available anymore, please check \'Scheduling\' tab.',
    ],
    'test' => [
        'test_comment' => 'Var god testa din ompekning innan du sparar den',
        'input_path' => 'Ange sökväg',
        'input_path_comment' => 'Sökvägen du vill testa. Ex. /old-blog/category/123',
        'input_path_placeholder' => '/input/path',
        'input_scheme' => 'Input scheme',
        'test_date' => 'Testdatum',
        'test_date_comment' => 'Om du har schemalagt denna ompekning kan du testa det här',
        'testing' => 'Testar...',
        'run_test' => 'Kör test',
        'no_match_label' => 'Sorry, ingen matchning!',
        'no_match' => 'Ingen matchning hittades!',
        'match_success_label' => 'Vi har en matchning!',
    ],
    'test_lab' => [
        'section_test_lab_comment' => 'TestLab allows you to mass test your redirects.',
        'test_lab_label' => 'Include in TestLab',
        'test_lab_enable' => 'Flick this switch to allow testing this redirect in the TestLab.',
        'test_lab_path_label' => 'Test Path',
        'test_lab_path_comment' => 'This path will be used when performing tests. Replace placeholders with real values.',
        'start_tests' => 'Start Tests',
        'start_tests_description' => 'Press the \'Start tests\' button to begin.',
        'edit' => 'Edit',
        'exclude' => 'Exclude',
        'exclude_confirm' => 'Are you sure want to exclude this redirect from TestLab?',
        'exclude_indicator' => 'Excluding redirect from TestLab',
        're_run' => 'Re-run',
        're_run_indicator' => 'Running tests, please wait...',
        'loop' => 'Loop',
        'match' => 'Match',
        'response_http_code' => 'Response HTTP code',
        'response_http_code_should_be' => 'Response HTTP code should be one of:',
        'redirect_count' => 'Redirect count',
        'final_destination' => 'Final Destination',
        'no_redirects' => 'No redirects have been marked with TestLab enabled. Please enable the option \'Include in TestLab\' when editing a redirect.',
        'test_error' => 'An error occurred when testing this redirect.',
        'flash_test_executed' => 'Test has been executed.',
        'flash_redirect_excluded' => 'Redirect has been excluded from TestLab and will not show up on next test run.',
        'result_request_failed' => 'Could not execute request.',
        'redirects_followed' => 'Number of redirects followed: :count (limited to :limit)',
        'not_determinate_destination_url' => 'Could not determine final destination URL.',
        'no_destination_url' => 'No final destination URL.',
        'final_destination_is' => 'Final destination is: :destination',
        'possible_loop' => 'Possible redirect loop!',
        'no_loop' => 'No redirect loop detected.',
        'not_match_redirect' => 'Did not match any redirect.',
        'matched' => 'Matched',
        'redirect' => 'redirect',
        'matched_not_http_code' => 'Matched redirect, but response HTTP code did not match! Expected :expected but received :received.',
        'matched_http_code' => 'Matched redirect, response HTTP code :code.',
        'executing_tests' => 'Executing tests...',
    ],
    'statistics' => [
        'hits_per_day' => 'Redirect hits per day',
        'click_on_chart' => 'Click on the chart to enable zooming and dragging.',
        'requests_redirected' => 'Requests redirected',
        'all_time' => 'all time',
        'active_redirects' => 'Active redirects',
        'redirects_this_month' => 'Redirects this month',
        'previous_month' => 'previous month',
        'latest_redirected_requests' => 'Latest redirected request',
        'redirects_per_month' => 'Redirects per month',
        'no_data' => 'No data',
        'top_crawlers_this_month' => 'Top :top crawlers this month',
        'top_redirects_this_month' => 'Top :top redirects this month',
        'activity_last_three_months' => 'Activity last 3 months',
    ],
    'title' => [
        'import' => 'Importera',
        'export' => 'Exportera',
        'redirects' => 'Hantera ompekningar',
        'create_redirect' => 'Skapa ompekning',
        'edit_redirect' => 'Ändra ompekning',
        'categories' => 'Hantera kategorier',
        'create_category' => 'Skapa kategori',
        'edit_category' => 'Ändra kategori',
        'view_redirect_log' => 'View redirect log',
        'statistics' => 'Statistics',
        'test_lab' => 'TestLab (beta)',
    ],
    'buttons' => [
        'add' => 'Skapa',
        'from_request_log' => 'Från anropslogg',
        'new_redirect' => 'Ny ompekning',
        'create_redirects' => 'Skapa ompekning',
        'create_redirect' => 'Create redirect',
        'create_and_new' => 'Create and new',
        'delete' => 'Radera',
        'enable' => 'Aktivera',
        'disable' => 'Inaktivera',
        'reorder_redirects' => 'Omsortera',
        'export' => 'Exportera',
        'import' => 'Importera',
        'settings' => 'Settings',
        'categories' => 'Kategorier',
        'extensions' => 'Extensions',
        'new_category' => 'Ny kategori',
        'reset_statistics' => 'Återställ statistik',
        'logs' => 'Redirect log',
        'empty_redirect_log' => 'Empty redirect log',
        'clear_cache' => 'Clear cache',
        'stop' => 'Stop',
        'reset_all' => 'Reset statistics for all redirects',
        'all_redirects' => 'all redirects',
        'bulk_actions' => 'Bulk actions',
    ],
    'tab' => [
        'tab_general' => 'Allmänt',
        'tab_requirements' => 'Krav',
        'tab_test' => 'Test',
        'tab_scheduling' => 'Schemalägg',
        'tab_test_lab' => 'TestLab',
        'tab_advanced' => 'Advanced',
        'tab_logs' => 'Event log',
    ],
    'flash' => [
        'success_created_redirects' => 'Skapade :count ompekningar',
        'static_page_redirect_not_supported' => 'This redirect cannot be modified. Plugin Winter.Pages is required.',
        'truncate_success' => 'Successfully deleted all records',
        'delete_selected_success' => 'Successfully deleted selected records',
        'cache_cleared_success' => 'Successfully cleared redirect cache',
        'statistics_reset_success' => 'All statistics have been successfully reset',
        'enabled_all_redirects_success' => 'All redirects have been successfully enabled',
        'disabled_all_redirects_success' => 'All redirects have been successfully disabled',
        'deleted_all_redirects_success' => 'All redirects have been successfully deleted',
    ],
    'import_export' => [
        'match_type' => 'Match Type [match_type] (Allowed values: exact, placeholders, regex)',
        'category_id' => 'Category [category_id]',
        'target_type' => 'Target Type [target_type] (Allowed values: path_or_url, cms_page, static_page, none)',
        'from_url' => 'Source path [from_url]',
        'from_scheme' => 'Source scheme [from_scheme] (Allowed values: http, https, auto [default])',
        'to_url' => 'Target path [to_url]',
        'to_scheme' => 'Target scheme [to_scheme] (Allowed values: http, https, auto [default])',
        'test_url' => 'Test URL [test_url]',
        'cms_page' => 'CMS Page [cms_page] (Filename without .htm extension)',
        'static_page' => 'Static Page [static_page] (Filename without .htm extension)',
        'requirements' => 'Placeholder requirements [requirements]',
        'status_code' => 'HTTP status code [status_code] (Possible values: 301, 302, 303, 404, 410)',
        'hits' => 'Redirect Hits [hits]',
        'from_date' => 'Scheduled date from [from_date] (YYYY-MM-DD or empty)',
        'to_date' => 'Scheduled date to [to_date] (YYYY-MM-DD or empty)',
        'sort_order' => 'Priority [sort_order]',
        'is_enabled' => 'Enabled [is_enabled] (1 = enable redirect, 0 = disable redirect [default])',
        'ignore_query_parameters' => 'Ignore Query Parameters [ignore_query_parameters] (1 = yes, 0 = no [default])',
        'ignore_case' => 'Ignore Case [ignore_case] (1 = yes, 0 = no [default])',
        'ignore_trailing_slash' => 'Ignore Trailing Slashes [ignore_trailing_slash] (1 = yes, 0 = no [default])',
        'test_lab' => 'Test Lab [test_lab] (1 = enable Test Lab, 0 = disable TestLab [default])',
        'test_lab_path' => 'Test Lab path [test_lab_path] (required if match_type = placeholders)',
        'system' => 'System [system] (1 = system generated redirect, 0 = user generated redirect [default])',
        'description' => 'Description [description]',
        'last_used_at' => 'Last Used At [last_used_at] (YYYY-MM-DD HH:MM:SS or empty)',
        'created_at' => 'Created At [created_at] (YYYY-MM-DD HH:MM:SS or empty)',
        'updated_at' => 'Updated At [updated_at] (YYYY-MM-DD HH:MM:SS or empty)',
    ],
];
